[IMP] mass_mailing_custom_unsubscribe: GDPR compliance

yajo opened the pull request:
- Record resubscriptions too.
- Record action metadata.

@Tecnativa

yajo committed:
[IMP] mass_mailing_custom_unsubscribe: GDPR compliance

- Record resubscriptions too.
- Record action metadata.

pedrobaeza commented on mass_mailing_custom_unsubscribe/__manifest__.py:
```diff
@@ -3,9 +3,9 @@
+    "summary": "Know and track (un)subscription reasons, GDPR compliant",
     'category': 'Marketing',
-    'version': '10.0.1.0.0',
+    'version': '10.0.1.1.0',
```
This deserves 10.0.2.0.0

yajo committed:
!fixup [IMP] mass_mailing_custom_unsubscribe: GDPR compliance

rafaelbn commented:
@Yajo I cannot tested in runbot because (see screenshot)

![2018-05-05_8-58-38](https://user-images.githubusercontent.com/8736623/39660600-a4eef388-5042-11e8-8e93-3a258d6a1e79.png)

yajo commented:
Lol, that seems like a misconfiguration in runbot's nginx... I'll test locally.

yajo committed:
!squash [IMP] mass_mailing_custom_unsubscribe: GDPR compliance

- Never require details in frontend if unsubscription form is hidden.
- Make ESLint happy.
- Quick color-based action distinction in tree view.
- Add useful quick groupings.
- Display (un)subscription metadata.

yajo commented:
I double-checked this works fine, and added some UX enhacements too.

chienandalu reviewed (changes requested):
Testing locally:
- I tried to unsubscribe with no reason set, and nothing stopped me from it.
- I unsubscribed and then subscribed again. Then unsubscribing seemed to be not possible from here

yajo committed:
!squash [IMP] mass_mailing_custom_unsubscribe: GDPR compliance

- Do not store duplicated (un)subscriptions.
- Require reasons for unsubscriptions only.
- Pivot & graph views.

yajo commented:
The new commit fixes the problems you detected, and besides it adds graph and pivot views for unsubscriptions.

rafaelbn commented:
@Yajo travis is :red_circle:

yajo committed:
fixup! !squash [IMP] mass_mailing_custom_unsubscribe: GDPR compliance

yajo committed:
fixup! !squash [IMP] mass_mailing_custom_unsubscribe: GDPR compliance

yajo committed:
fixup! !squash [IMP] mass_mailing_custom_unsubscribe: GDPR compliance

yajo committed:
fixup! !squash [IMP] mass_mailing_custom_unsubscribe: GDPR compliance

yajo commented:
It should go :green_heart: now

chienandalu reviewed (commented):
Just a little flaw now:

- Given a contact in two lists:
  1. Unsubscribe from `list1`
  2. Unsubscribe from `list2`and resubscribe to `list1`

Everything goes fine except the resubscription log to `list1` gets the unsubscription from `list2`reason.

yajo commented:
Good catch, it should be fixed now.

chienandalu reviewed (commented):
Now it won't let me unsubscribe: **Your changes have not been saved, try again later**

chienandalu commented:
A trace here:

<details>

```
odoo_1  | 2018-05-08 12:17:59,655 45 ERROR test odoo.http: Exception during JSON request handling.
odoo_1  | Traceback (most recent call last):
odoo_1  |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 641, in _handle_exception
odoo_1  |     return super(JsonRequest, self)._handle_exception(exception)
odoo_1  |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 683, in dispatch
odoo_1  |     result = self._call_function(**self.params)
odoo_1  |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 333, in _call_function
odoo_1  |     return checked_call(self.db, *args, **kwargs)
odoo_1  |   File "/opt/odoo/custom/src/odoo/odoo/service/model.py", line 101, in wrapper
odoo_1  |     return f(dbname, *args, **kwargs)
odoo_1  |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 326, in checked_call
odoo_1  |     result = self.endpoint(*a, **kw)
odoo_1  |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 941, in __call__
odoo_1  |     return self.method(*args, **kw)
odoo_1  |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 506, in response_wrap
odoo_1  |     response = f(*args, **kw)
odoo_1  |   File "/opt/odoo/auto/addons/mass_mailing_custom_unsubscribe/controllers/main.py", line 113, in unsubscribe
odoo_1  |     mailing._unsubscribe_token(res_id, token)
odoo_1  |   File "/opt/odoo/auto/addons/mass_mailing_custom_unsubscribe/models/mail_mass_mailing.py", line 34, in _unsubscribe_token
odoo_1  |     key = (self.env.cr.dbname, self.id, int(res_id))
odoo_1  | TypeError: int() argument must be a string or a number, not 'NoneType'
```

</details>

yajo commented:
That's not supposed to happen. It's like you're using a wrong URL to unsubscribe. Are you testing the right way?

chienandalu commented:
I'm testing the same way as before:

- create a new mass mailing
- send this mass_mailing to a contact in two lists
- open the link in incognito window
- try to unsubscribe

yajo commented:
Make sure you test with real emails, not with test emails. Test email unsubscription shouldn't work.

chienandalu reviewed (approved):
Works great now!
